tests/helpers/file-queue-helper-test.js

```diff
@@ -0,0 +1,162 @@
+    ]);
+  });
+
+  test('falls back to default name', async function (assert) {
+    await render(hbs`
+      {{#let (file-queue) as |queue|}}
+        <output>{{queue.name}}</output>
+      {{/let}}
+    `);
+
+    assert.dom('output').hasText(DEFAULT_QUEUE.toString());
+  });
+
+  test('can be parametrized by name', async function (assert) {
+    await render(hbs`
+      {{#let (file-queue name="line-up") as |queue|}}
+        <output>{{queue.name}}</output>
+      {{/let}}
+    `);
+
+    assert.dom('output').hasText('line-up');
+  });
+
+  test('will be notified when adding and removing files', async function (assert) {
+    this.addFile = (file) => assert.step(`file added: ${file.name}`);
+    this.removeFile = (file) => assert.step(`file removed: ${file.name}`);
+
+    await render(hbs`
+      {{#let (file-queue fileAdded=this.addFile fileRemoved=this.removeFile) as |queue|}}
+        {{#each queue.files as |file|}}
+          <span data-test-file>
+          {{file.name}}
+          </span>
+          <button
+            type="button"
+            data-test-remove
+            {{on "click" (fn queue.remove file)}}
+          >
+            x
+          </button>
+        {{/each}}
+
+        <label>
+          <input type="file" {{queue.selectFile}} hidden>
+          Select File
+        </label>
+      {{/let}}
+    `);
+
+    assert.dom('[data-test-file]').doesNotExist();
+
+    await selectFiles('input[type=file]', new File([], 'dingus.txt'));
+    assert.dom('[data-test-file]').hasText('dingus.txt');
+
+    await click('[data-test-remove]');
+    assert.verifySteps(['file added: dingus.txt', 'file removed: dingus.txt']);
+
+    assert.dom('[data-test-file]').doesNotExist();
+  });
+
+  test('uploading twice will manage files in queue', async function (assert) {
+    this.server.post(
+      '/folder/:id/file',
+      uploadHandler(function (/*schema, request*/) {
+        // do sth
+      })
+    );
+
+    this.uploadImage = (file) => {
+      return file.upload('/folder/1/file');
+    };
+
+    await render(hbs`
+      {{#let (file-queue fileAdded=this.uploadImage) as |queue|}}
+        {{#each queue.files as |file|}}
+          <span data-test-file>
+          {{file.name}}
+          </span>
+        {{/each}}
+
+        <label>
+          <input type="file" {{queue.selectFile}} hidden>
+          Select File
+        </label>
+      {{/let}}
+    `);
+
+    assert.dom('[data-test-file]').doesNotExist();
+
+    // first upload
+    selectFiles('input[type=file]', new File([], 'dingus.txt'));
+    await waitFor('[data-test-file]');
+    assert.dom('[data-test-file]').hasText('dingus.txt');
+
+    await settled();
+    assert.dom('[data-test-file]').doesNotExist();
+
+    // second upload (retention teset)
+    // to check `Queue.flush()` will keep auto-tracking intact
+    selectFiles('input[type=file]', new File([], 'dingus.txt'));
+    await waitFor('[data-test-file]');
+    assert.dom('[data-test-file]').hasText('dingus.txt');
```

Comment:
@gossi Would writing `await selectFiles(...)` and removing `await waitFor(...)` help with the failing test in CI?

Reply:
I see what you're trying to do here – but even locally for me the timing of this test seems unreliable.

When a file is selected and immediately uploaded the `Queue.flush` method should run, but it all happens in a very short time.

Suspect glimmer is optimising for minimal renders and deciding not to update the DOM with every change here. I got it to pass more often by adjusting the timing of mirage but it still wasn't and acceptable failure rate.

Could you explain clearly what you're trying to test? You're checking rendered page state since you mention autotracking, yeah? Is there a simpler way to test this?

Reply:
Well, I wanna test the "uploading" state and that the UI is reacting to that state change.

When doing `await selectFiles()` this will be skipping over the `uploading` state to whatever comes after. So, not awaiting this and `waitFor()` the element that should pick up the UI change kinda seemed the way to go - and for what it's worth, it worked for the situations I used that. I also had tests, that weren't possible this way.

I also used a waiting timer in the BE (I used polly with a delay in our app code), that didn't worked at all, the `waitFor()` was much more reliable than the timer. My very assumptive exaplanation is that the runloop is queuing those steps in and allows testing in that order, if they are very laboratory and an execution in this order can be guaranteed.

That said, I'm happily taking advice for this one. Maybe we can provide better test adapters? (for mirage and polly?)

Reply:
I've used a timeout for this case here. Let's see if CI is happy with that. Also this is not a problem introduced as part of this PR, more that this PR uncovered this issue.

As a matter of that, I created an issue to discuss ideas for solving that in the future #641 and keep the scope of this PR to "make-the-CI-green"